mne/beamformer/_ap.py

```diff
@@ -0,0 +1,765 @@
+    nsources : int
+        The number of dipoles to estimate.
+    picks : List of int
+        Channel indiecs for filtering.
+    max_iter : int
+        Maximal iteration number of AP.
+    force_no_rep : bool
+        Forces no repetition of estimated dipoles.
+
+    Returns
+    -------
+    dipoles : list of instances of Dipole
+        The dipole fits.
+    explained_data : array
+        Data explained by the dipoles using a least square fitting with the
+        selected active dipoles and their estimated orientation.
+    var_exp : float
+        Percentile of data variation explained (see: _log_exp_var).
+    dip_ind : List of int
+        List of indices of dipole source estimated.
+    oris : array, shape (nsources, 3)
+        Orientations array of estimated sources (sorted by dip_ind).
+    poss : array, shape (nsources, 3)
+        Coordinates array of estimated sources (sorted by dip_ind).
+
+    """
+    info = pick_info(info, picks)
+    del picks
+
+    if forward['surf_ori'] and not is_fixed_orient(forward):
+        forward = convert_forward_solution(forward, surf_ori=False)
+    is_free_ori, info, _, _, gain, whitener, _, _ = _prepare_beamformer_input(
+        info, forward, noise_cov=noise_cov, rank=None)
+    forward = pick_channels_forward(forward, info['ch_names'], ordered=True)
+    del info
+
+    # whiten the data (leadfield already whitened)
+    wh_data = np.dot(whitener, data)
+    del data
+
+    if is_free_ori:
+        idx, oris, poss, gain_active, gain_dip, sol, dip_ind = \
+            _free_ori_ap(wh_data, gain, nsources, forward,
+                         max_iter=max_iter, force_no_rep=force_no_rep)
+        X = sol[:, np.newaxis] * oris[:, :, np.newaxis]
+        X.shape = (-1, len(times))
+    else:
+        idx, oris, poss, gain_active, gain_dip, sol = \
+            _fixed_ori_ap(wh_data, gain, nsources, forward,
+                          max_iter=max_iter)
+        X = sol
+        dip_ind = idx
+
+    gain_active = whitener @ gain_active.reshape(gain.shape[0], -1)
+    explained_data = gain_dip @ sol
+    m_estimate = whitener @ explained_data
+    var_exp = _log_exp_var(wh_data, m_estimate)
+    tstep = np.median(np.diff(times)) if len(times) > 1 else 1.
+    dipoles = _make_dipoles_sparse(
+        X, idx, forward, times[0], tstep, wh_data,
+        gain_active, active_is_idx=True)
+    for dipole, ori in zip(dipoles, oris):
+        signs = np.sign((dipole.ori * ori).sum(-1, keepdims=True))
+        dipole.ori *= signs
+        dipole.amplitude *= signs[:, 0]
+    logger.info('[done]')
+
+    return dipoles, explained_data, var_exp, dip_ind, oris, poss
+
+
+def _residual_packing(evoked, picks, explained_data_mat, info):
+    """Pack the residual data into an mne.Evoked object."""
+    residual = evoked.copy()
+    selection = [info['ch_names'][pick] for pick in picks]
+
+    residual = pick_channels_evoked(residual,
+                                    include=selection)
```

Comment:
nowadays this would usually be something like
```suggestion
    residual = residual.pick(picks)
```

(but please double-check this gives the result you expect... `pick_channels_evoked` is very old code so I'm not intimately familiar with its quirks)

Reply:
updated the method used